[MemProf] Re-enable cloning of callsites in recursive cycles with fixes

teresajohnson opened the pull request:
This change addresses a number of issues with the support added by
PR121985 which were exposed through more exhaustive testing,
specifically places that needed updates to perform correct graph updates
in the presence of cycles.

A new test case is added that reproduces these issues, and the default
is flipped back to enabling this handling.

teresajohnson committed:
[MemProf] Re-enable cloning of callsites in recursive cycles with fixes

This change addresses a number of issues with the support added by
PR121985 which were exposed through more exhaustive testing,
specifically places that needed updates to perform correct graph updates
in the presence of cycles.

A new test case is added that reproduces these issues, and the default
is flipped back to enabling this handling.

snehasish commented on llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:
```diff
@@ -3946,6 +4027,11 @@ bool CallsiteContextGraph<DerivedCCG, FuncTy, CallTy>::assignFunctions() {
         // handling and makes it less error-prone.
         auto CloneCallerEdges = Clone->CallerEdges;
         for (auto &Edge : CloneCallerEdges) {
+          // Skip removed edges (due to direct recursive edges updated when
```
unclosed parenthesis?

teresajohnson replied:
fixed

snehasish reviewed (approved):
LGTM. I think we need to update this [test](http://github.com/llvm/llvm-project/blob/main/llvm/test/Transforms/MemProfContextDisambiguation/recursive.ll) ([buildkite](https://buildkite.com/llvm-project/github-pull-requests/builds/144058)).

teresajohnson committed:
Fix comment and update test

teresajohnson commented:
> LGTM. I think we need to update this [test](https://github.com/llvm/llvm-project/blob/main/llvm/test/Transforms/MemProfContextDisambiguation/recursive.ll) ([buildkite](https://buildkite.com/llvm-project/github-pull-requests/builds/144058)).

Woops, updated the other version of the recursive test, now fixed this one the same way
